instrumentation/github.com/gofiber/fiber/otelfiber/fiber.go

```diff
@@ -0,0 +1,123 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otelfiber
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/gofiber/fiber/v2"
+
+	otelcontrib "go.opentelemetry.io/contrib"
+	"go.opentelemetry.io/otel"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/propagation"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	oteltrace "go.opentelemetry.io/otel/trace"
+)
+
+const (
+	tracerKey  = "otel-go-contrib-tracer-gofiber-fiber"
+	tracerName = "go.opentelemetry.io/contrib/instrumentation/github.com/gofiber/fiber/otelfiber"
+)
+
+// Middleware returns fiber handler which will trace incoming requests.
+func Middleware(service string, opts ...Option) fiber.Handler {
+	cfg := config{}
+	for _, opt := range opts {
+		opt.apply(&cfg)
+	}
+	if cfg.TracerProvider == nil {
+		cfg.TracerProvider = otel.GetTracerProvider()
+	}
+	tracer := cfg.TracerProvider.Tracer(
+		tracerName,
+		oteltrace.WithInstrumentationVersion(otelcontrib.SemVersion()),
+	)
+	if cfg.Propagators == nil {
+		cfg.Propagators = otel.GetTextMapPropagator()
+	}
+
+	return func(c *fiber.Ctx) error {
+		c.Locals(tracerKey, tracer)
+		savedCtx, cancel := context.WithCancel(c.Context())
+
+		defer func() {
+			c.SetUserContext(savedCtx)
+			cancel()
+		}()
+
+		reqHeader := make(http.Header)
+		c.Request().Header.VisitAll(func(k, v []byte) {
+			sk := string(k)
+			sv := string(v)
+
+			switch sk {
+			case fiber.HeaderTransferEncoding:
+				reqHeader[fiber.HeaderTransferEncoding] = append(reqHeader[fiber.HeaderTransferEncoding], sv)
+			default:
+				reqHeader[sk] = []string{sv}
```

Comment:
I am not sure that this handling is correct. Here is an example why: https://www.w3.org/TR/trace-context/#tracestate-header-field-values

> Tracestate MAY be sent or received as multiple header fields. Multiple tracestate header fields MUST be handled as specified by RFC7230 Section 3.2.2 Field Order. The tracestate header SHOULD be sent as a single field when possible, but MAY be split into multiple header fields.

Reply:
hmmm i borrow the code from fasthttp

https://github.com/valyala/fasthttp/blob/b3ece39dcdd40ab814988d7a322b2f1a050e13f4/fasthttpadaptor/request.go#L46-L56

Reply:
something like this instead? 

```go
c.Request().Header.VisitAll(func(k, v []byte) {
	sk := string(k)
	sv := string(v)
	reqHeader.Add(sk, sv)
})
```